Sources/CacheAdvance/FileHeader.swift

```diff
@@ -0,0 +1,150 @@
+            return nil
+        }
+
+        let versionData = data.subdata(in: FileHeader.Field.version.rangeOfFieldInFile)
+        let maximumBytesData = data.subdata(in: FileHeader.Field.maximumBytes.rangeOfFieldInFile)
+        let overwritesOldMessagesData = data.subdata(in: FileHeader.Field.overwriteOldMessages.rangeOfFieldInFile)
+        let offsetInFileOfOldestMessageData = data.subdata(in: FileHeader.Field.offsetInFileOfOldestMessage.rangeOfFieldInFile)
+        let offsetInFileAtEndOfNewestMessageData = data.subdata(in: FileHeader.Field.offsetInFileAtEndOfNewestMessage.rangeOfFieldInFile)
+
+        let version = UInt8(versionData)
+        let maximumBytes = Bytes(maximumBytesData)
+        let overwritesOldMessages = Bool(overwritesOldMessagesData)
+        let offsetInFileOfOldestMessage = UInt64(offsetInFileOfOldestMessageData)
+        let offsetInFileAtEndOfNewestMessage = UInt64(offsetInFileAtEndOfNewestMessageData)
+
+        self = FileHeader(
+            version: version,
+            maximumBytes: maximumBytes,
+            overwritesOldMessages: overwritesOldMessages,
+            offsetInFileOfOldestMessage: offsetInFileOfOldestMessage,
+            offsetInFileAtEndOfNewestMessage: offsetInFileAtEndOfNewestMessage)
+    }
+
+    // MARK: Internal
+
+    let version: UInt8
+    let maximumBytes: Bytes
+    let overwritesOldMessages: Bool
+    let offsetInFileOfOldestMessage: UInt64
+    let offsetInFileAtEndOfNewestMessage: UInt64
+
+    var asData: Data {
+        Field.allCases.reduce(Data()) { header, field in
+            header + data(for: field)
+        }
+    }
+
+    /// The expected version of the header.
+    /// Whenever there is a breaking change to the header format, update this value.
+    static let version: UInt8 = 1
+
+    /// Calculates the offset in the file where the header should end.
+    static var expectedEndOfHeaderInFile = Field.endOfHeaderMarker.expectedEndOfFieldInFile
+
+    func data(for field: Field) -> Data {
+        switch field {
+        case .version:
+            return Data(version)
+        case .maximumBytes:
+            return Data(maximumBytes)
+        case .overwriteOldMessages:
+            return Data(overwritesOldMessages)
+        case .offsetInFileOfOldestMessage:
+            return Data(offsetInFileOfOldestMessage)
+        case .offsetInFileAtEndOfNewestMessage:
+            return Data(offsetInFileAtEndOfNewestMessage)
+        case .reservedSpace:
+            return Data(repeating: 0, count: field.storageLength)
+        case .endOfHeaderMarker:
+            return Data()
+        }
+    }
+
+    enum Field: Int, CaseIterable {
+        // Header format:
+        // [headerVersion:UInt8][maximumBytes:UInt64][overwritesOldMessages:Bool][offsetInFileOfOldestMessage:UInt64][offsetInFileAtEndOfNewestMessage:UInt64][reservedSpace]
+
+        case version = 0
+        case maximumBytes
+        case overwriteOldMessages
+        case offsetInFileOfOldestMessage
```

Comment:
Note that we store this value even in CacheAdvances that do not `overwriteOldMessages`, despite the fact that this value will always be equal to `expectedEndOfHeaderInFile` when a cache doesn't overwrite old messages.

Having a consistent header between the different CacheAdvance types has the benefit of limiting complexity.